### ISSUE TITLE
Proposal: minimal changes to allow XDG-compatible data storage.

### ISSUE THREAD

**Issue by lgarron** [microsoft/vscode#162712]
There are many of us who find it bad practice that VSCode creates a `~/.vscode` folder instead of following the common [XDG base directory structure](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), respected by a wide variety of programs. This issue has come up regularly, without a good permanent solution:

- https://github.com/microsoft/vscode/issues/3884
- https://github.com/microsoft/vscode/issues/84808
- https://github.com/microsoft/vscode/issues/100988
- https://github.com/microsoft/vscode/issues/17691
- https://github.com/microsoft/vscode/issues/126182

I would like to propose the following new behaviour:

1. If `~/.config/[vs]code[-insiders]/argv.json` is present at launch in non-portable mode, then VSCode uses this as the path for `argv.json`.
  - The brackets denote prefix and suffix of the installed build, per https://github.com/microsoft/vscode/issues/3884 . This is deterministic for any given installed build, and would be `~/.config/vscode/argv.json` for "most" people.
2. `argv.json` supports setting the following arguments, which the user can set to XDG-compatible paths if they like:
  - `user-data-dir`
  - `extensions-dir`

As far as I can tell, this would be sufficient to avoid creating a `~/.vscode` folder. It also reuses the existing structure of VSCode to do this without introducing significant hacks, edge cases, or race conditions.

Part 1 seems reasonably feasible to me, given that `argv.json` is hardcoded in delightfully few places at the moment: https://github.com/microsoft/vscode/search?q=%22argv.json%22&type=code

Part 2 would essentially involve adding those arguments to [`SUPPORTED_ELECTRON_SWITCHES`](https://github.com/microsoft/vscode/blob/9db57e76e9f16f552203585a394e50cf0f7a84a7/src/main.js#L153).

I'd be happy to submit a PR if this kind of approach is amenable to the maintainers.

**Comment by Da-Viper** [comment]
@bpasero I want to try this issue as it is marked as good first issue, 

From what I understand this is to make the `argv.json` file comply with platform specific  appdata path. In extension the XDG paths for linux ?

**Comment by bpasero** [comment]
Yes, basically instead of just using `os.homedir`, support to check for presence of XDG environment variable. We probably also need to migrate the file to not break existing users.

**Comment by Da-Viper** [comment]
> Yes, basically instead of just using `os.homedir`, support to check for presence of XDG environment variable. We probably also need to migrate the file to not break existing users.

This is what I assume this would be the Migration style: 


**Scenario 1** 
Exists: oldpath, newpath. 
Resolution: Delete the old file   // not sure if i am to merge the two files together ?? 
return the new path

**Scenario 2** 
Exists: oldpath. 
Resolution: Move the old path to the new path
return the new path

**Scenario 3**
Exists: newpath 
Resolution: Nothing
return the new path

**Scenario 4**
Exists: None
Resolution: Nothing    //creating of new argv.json logic is handled else where 
return the new path

**Comment by untainsYD** [comment]
Is there any solution for this `ISSUE`? Because I have been trying to create `~/.config/vscode` folder with `user-data-dir` and `extensions-dir` paths configured in my `~/.config/code-flags.conf` but `~/.vscode` folder is still created by VSCode startup in my `$HOME` directory.

**Comment by gjsjohnmurray** [comment]
@untainsYD the PR #170364 by @Da-Viper hasn't yet been merged. In fact it seems to have failed a CI action so may need attention by the author before it is ready for @bpasero or another member of the core team to review.

**Comment by Da-Viper** [comment]
> @untainsYD the PR #170364 by @Da-Viper hasn't yet been merged. In fact it seems to have failed a CI action so may need attention by the author before it is ready for @bpasero or another member of the core team to review.

I tried to update the MR but it is requesting review before CI

**Comment by untainsYD** [comment]
Any updates on that ISSUE?

**Comment by ashutosh887** [comment]
I'm willing to work on this

**Comment by untainsYD** [comment]
> > @untainsYD the PR #170364 by @Da-Viper hasn't yet been merged. In fact it seems to have failed a CI action so may need attention by the author before it is ready for @bpasero or another member of the core team to review.
> 
> I tried to update the MR but it is requesting review before CI


In this issue `~/.pki` folder should be also taken to the account.
> VSCode also generates `~/.pki ` when the vanilla electron stops doing it.

**Comment by sdavids** [comment]
While you are at it, please do not forget about macOS:

https://github.com/microsoft/vscode/blob/7ca99004a50625858c0c2c0dbc2a9845e41a7dea/src/vs/platform/environment/node/userDataPath.js#L91-L92

**Comment by ispanos** [comment]
I don't know if this should be added in this issue, but the R extension also creates `.vscode-R` in the home directory. In total I have 3 hidden directories, `.pki`, '.vscode`, `vscoode-R`, just from code.

**Comment by DAWSpence** [comment]
I would like to work on this issue.  What are the policies of claiming an issue such as this?

**Comment by untainsYD** [comment]
Now it creates `.dotnet` folder. Are there any updates on this annoying issue?

**Comment by doraeric** [comment]
Microsoft is not fixing the issue, and is even adding new folders to the home directory.
The recent PR #311317 writes to `~/.vscode-shared`. I have blocked VS Code from creating files in my home directory. However, since version 1.118.0, VS Code cannot even load recently opened folders on startup because it is blocked from creating this folder.

@sandy081 Could you please revert that PR or follow XDG spec?

**Comment by lgarron** [comment]
> Microsoft is not fixing the issue, and is even adding new folders to the home directory. The recent PR [#311317](https://github.com/microsoft/vscode/pull/311317) writes to `~/.vscode-shared`. I have blocked VS Code from creating files in my home directory. However, since version 1.118.0, VS Code cannot even load recently opened folders on startup because it is blocked from creating this folder.
> 
> [@sandy081](https://github.com/sandy081) Could you please revert that PR or follow XDG spec?

@bpasero @roblourens @alexr00 @dmitrivMS (reviewers of the PR): Would you be open to either:

- reverting https://github.com/microsoft/vscode/pull/311317 or
- accepting any PR that changes this not to create a new folder under `~/` by default,

and specifically in a timeframe on the order of a week or two?

I know it's not very insightful to say this, but a lot of us feel like we are shouting into the void here. I'd be willing to put in the time into a PR that matches any backwards compatibility requirements you communicate.

**Comment by dmitrivMS** [comment]
I was hoping to look at some of these issues, but I can't promise it will happen in next week.
Not creating folder until necessary sounds like a good mitigation - could you please link that PR?

**Comment by lgarron** [comment]
> I was hoping to look at some of these issues, but I can't promise it will happen in next week.

Thanks for the reply!

> Not creating folder until necessary sounds like a good mitigation - could you please link that PR?

Unless I'm missing something, this would not be a particularly good mitigation.

- The new location is used for workspace trust, and
- workspace trust is needed to open a shell.

I don't know who develops without opening a shell, but that certainly wouldn't describe me.

(I also don't know what PR I might be linking to in this case.)

**Comment by dmitrivMS** [comment]
Ah, sorry, I misread - I thought it said "my PR", but you asked if there was a PR. Well, then we need to create one...

**Comment by TROLlox78** [comment]
We are asking if you could please try to follow the XDG spec which says that you should store your configuration files in a path provided by the environment variable `XFG_CONFIG_HOME` or otherwise default to storing in `~/.config`
Thank you

**Comment by dmitrivMS** [comment]
@TROLlox78 Yes, thank you, I understand that. It's a larger change though, so I was wondering if there's an issue we can address which would help, but is smaller. Perhaps there is not...

**Comment by doraeric** [comment]
> [@TROLlox78](https://github.com/TROLlox78) Yes, thank you, I understand that. It's a larger change though, so I was wondering if there's an issue we can address which would help, but is smaller. Perhaps there is not...

@dmitrivMS, thanks for your attention. Starting with my [comment](https://github.com/microsoft/vscode/issues/162712#issuecomment-4358620158), we are talking about the specific folder `~/.vscode-shared` (`sharedDataFolderName`) introduced by PR #311317, which was merged 2 weeks ago.

I can copy the text from the comment to create a new issue. If that's what you need, just let me know and I will do so.

Since it was merged only 2 weeks ago, we can prevent @sandy081 from making the issue more serious. I believe it's not too late to fix it before we need a really **LARGE** change.

**Comment by dmitrivMS** [comment]
@doraeric I looked at making storage folder creation lazy and it feels little more extensive than I feel comfortable as a mitigation. If we go for a larger change, I'd rather try to tackle the ultimate goal which is XDG_CONFIG_HOME here. The reason you are hitting the issues is because recent files list is stored (has been moved) to the shared storage.

**Comment by levonk** [comment]
Related: opened #330061 for a similar XDG compliance gap — the `APPLICATION_SHARED` storage path (`~/.vscode-shared/sharedStorage/state.vscdb`) introduced in PR #311317. That path should default to `$XDG_STATE_HOME` (`~/.local/state/`) per the [XDG spec v0.8](https://specifications.freedesktop.org/basedir-spec/latest/).

Both this issue and #330061 are focused sub-issues under the #3884 umbrella, each addressing a specific path and XDG variable:
- **#162712** (this): `argv.json` → `$XDG_CONFIG_HOME`
- **#330061**: `sharedStorage/state.vscdb` → `$XDG_STATE_HOME`

The `state.vscdb` contents (recently-opened paths, trust markers) are state, not config — so `$XDG_STATE_HOME` is the correct category, distinct from the `$XDG_CONFIG_HOME` ask here.
